File: Assets/Scripts/Framework/Base/UIBase.cs
using System;
using Lite.Framework.Helper;
using UnityEngine;

namespace Lite.Framework.Base
{
    public enum UIDepthMode
    {
        Bottom = 0,
        Normal = 5000,
        Top = 10000
    }

    public class UIBase
    {
        public uint ID { get; }
        public string Name { get; set; }
        public string Path { get; set; }
        public Transform UITransform { get; set; }
        public RectTransform UIRectTransform { get; set; }
        public int Depth { get; protected set; }
        public UIDepthMode DepthMode { get; protected set; }

        public int SortOrder => Depth + (int)DepthMode;

        public UIBase()
        {
            ID = IDGenerator.Get();
            Name = GetType().Name;
            UITransform = null;
            UIRectTransform = null;
            Depth = 0;
            DepthMode = UIDepthMode.Normal;
        }

        public void Open(params object[] Params)
        {
            OnOpen(Params);
            Show();
        }

        public void Close()
        {
            Hide();
            UIHelper.RemoveAllEvent(UITransform, true);
            OnClose();
        }

        public void Show()
        {
            OnShow();
            UITransform.gameObject.SetActive(true);
        }

        public void Hide()
        {
            OnHide();
            UITransform.gameObject.SetActive(false);
        }

        public void Tick(float DeltaTime)
        {
            OnTick(DeltaTime);
        }

        public Transform FindChild(string ChildPath)
        {
            return UIHelper.FindChild(UITransform, ChildPath);
        }

        public T FindComponent<T>(string ChildPath) where T : Component
        {
            return UIHelper.FindComponent<T>(UITransform, ChildPath);
        }

        public Component FindComponent(string ChildParent, Type CType)
        {
            return UIHelper.FindComponent(UITransform, ChildParent, CType);
        }

        public Component FindComponent(string ChildParent, string CType)
        {
            return UIHelper.FindComponent(UITransform, ChildParent, CType);
        }

        public void AddEvent(Action<GameObject> Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.AddEvent(UITransform, Callback, Type);
        }

        public void RemoveEvent(Action<GameObject> Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.RemoveEvent(UITransform, Callback, Type);
        }

        public void AddEventToChild(string ChildPath, Action<GameObject> Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.AddEventToChild(UITransform, ChildPath, Callback, Type);
        }

        public void RemoveEventFromChild(string ChildPath, Action<GameObject> Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.RemoveEventFromChild(UITransform, ChildPath, Callback, Type);
        }

        /*public void AddEvent(Action Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.AddEvent(UITransform, Callback, Type);
        }

        public void RemoveEvent(Action Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.RemoveEvent(UITransform, Callback, Type);
        }

        public void AddEventToChild(string ChildPath, Action Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.AddEventToChild(UITransform, ChildPath, Callback, Type);
        }

        public void RemoveEventFromChild(string ChildPath, Action Callback, UIEventType Type = UIEventType.Click)
        {
            UIHelper.RemoveEventFromChild(UITransform, ChildPath, Callback, Type);
        }*/

        public void ShowChild(string ChildPath)
        {
            UIHelper.ShowChild(UITransform, ChildPath);
        }

        public void HideChild(string ChildPath)
        {
            UIHelper.HideChild(UITransform, ChildPath);
        }

        public void EnableTouched(bool Enabled)
        {
            UIHelper.EnableTouched(UITransform, Enabled);
        }

        public void EnableTouched(string ChildPath, bool Enabled)
        {
            UIHelper.EnableTouched(UITransform, ChildPath, Enabled);
        }

        protected virtual void OnOpen(params object[] Params)
        {
        }

        protected virtual void OnClose()
        {
        }

        protected virtual void OnShow()
        {
        }

        protected virtual void OnHide()
        {
        }

        protected virtual void OnTick(float DeltaTime)
        {
        }
    }
}

File: Assets/StandaloneAssets/scripts/framework/base/uibase.lua
UIBase = {}

--local _UIHelper = CS.Lite.Framework.UIHelper

function UIBase:FindChild(path)
	print(type(self._CSEntity_))
	return self._CSEntity_:FindChild(path)
end

function UIBase:FindComponent(path, ctype)
	return self._CSEntity_:FindComponent(path, ctype)
end

function UIBase:AddEvent(func, etype)
	etype = etype or UIEventType.Click
	self._CSEntity_:AddEvent(func, etype)
end

function UIBase:RemoveEvent(func, etype)
	etype = etype or UIEventType.Click
	self._CSEntity_:RemoveEvent(func, etype)
end

function UIBase:AddEventToChild(path, func, etype)
	etype = etype or UIEventType.Click
	self._CSEntity_:AddEventToChild(path, func, etype)
end

function UIBase:RemoveEventFromChild(path, func, etype)
	etype = etype or UIEventType.Click
	self._CSEntity_:RemoveEventFromChild(path, func, etype)
end

function UIBase:ShowChild(path)
	self._CSEntity_:ShowChild(path)
end

function UIBase:HideChild(path)
	self._CSEntity_:HideChild(path)
end

function UIBase:EnableTouched(enabled)
	self._CSEntity_:EnableTouched(enabled)
end

function UIBase:EnableTouched(path, enabled)
	self._CSEntity_:EnableTouched(path, enabled)
end

function UIBase:ExecuteMotion(motion)
	self._CSEntity_:ExecuteMotion(motion)
end

function UIBase:AbandonMotion(motion)
	self._CSEntity_:AbandonMotion(motion)
end